src/zjs_spi.c

```diff
@@ -0,0 +1,319 @@
+    handle->closed = true;
+    zjs_free(handle);
+}
+
+static const jerry_object_native_info_t spi_type_info =
+{
+   .free_cb = zjs_spi_callback_free
+};
+
+static ZJS_DECL_FUNC(zjs_spi_transceive)
+{
+    // requires: Writes and reads from the SPI bus, takes one argument
+    //           arg[0] - Buffer to write to the SPI bus
+    //  effects: Writes the buffer to the SPI bus and returns the buffer
+    //           received on the SPI in reply.
+
+    ZJS_VALIDATE_ARGS(Z_NUMBER, Z_OPTIONAL Z_ARRAY Z_STRING, Z_OPTIONAL Z_STRING);
+
+    // No more than 3 args
+    if (argc > 3)
+        return ZJS_STD_ERROR(RangeError, "Invalid args");
+
+    ZJS_GET_HANDLE(this, spi_handle_t, handle, spi_type_info);
+
+    if (handle->closed == true) {
+        ZJS_PRINT("SPI bus is closed\n");
+        return jerry_create_null();
+    }
+
+    size_t len;
+    jerry_value_t buffer;
+    zjs_buffer_t *tx_buf;
+    zjs_buffer_t *rx_buf;
+    jerry_value_t rx_buf_obj = jerry_create_null();
+    jerry_value_t tx_buf_obj;
+
+    if (spi_slave_select(handle->spi_device, jerry_get_number_value(argv[0]))) {
+        return zjs_error("SPI slave select failed\n");
+    }
+
+    // If only a slave is given, this must be a single read or its invalid
+    if (argc == 1) {
+        if (handle->topology != ZJS_TOPOLOGY_READ) {
+            return ZJS_STD_ERROR(RangeError, "Missing transmit buffer");
+        }
+    }
+
+    char dir_str[13];
+    // Set the direction default based on the topology.
+    enum spi_topology dir_arg = handle->topology;
+
+    // If we have a 'direction' arg, get it and validate
+    if (argc == 3) {
+        uint32_t dirLen = jerry_get_string_size(argv[2]) + 1;
+        zjs_copy_jstring(argv[2], dir_str, &dirLen);
+
+        if (strncmp(dir_str, "read-write", 11) == 0)
+            dir_arg = ZJS_TOPOLOGY_FULL_DUPLEX;
+        else if (strncmp(dir_str, "read", 5) == 0)
+            dir_arg = ZJS_TOPOLOGY_READ;
+        else if (strncmp(dir_str, "write", 6) == 0)
+            dir_arg = ZJS_TOPOLOGY_WRITE;
+        else
+            return ZJS_STD_ERROR(RangeError, "Invalid direction");
+
+        // If topology conflicts with direction given
+        if ((handle->topology == ZJS_TOPOLOGY_WRITE &&
+            dir_arg != ZJS_TOPOLOGY_WRITE) ||
+            (handle->topology == ZJS_TOPOLOGY_READ &&
+            dir_arg != ZJS_TOPOLOGY_READ)) {
+            return ZJS_STD_ERROR(NotSupportedError, "Direction conflicts with topology");
+        }
+        // If reading only, the 2nd arg should be NULL
+        if (dir_arg == ZJS_TOPOLOGY_READ && !jerry_value_is_null(argv[1])) {
+            return ZJS_STD_ERROR(NotSupportedError, "Buffer should be null when direction is read");
+        }
+    }
+
+    // If we need to write a buffer
+    if (dir_arg != ZJS_TOPOLOGY_READ) {
+        buffer = argv[1];
+        // Figure out if the buffer is an array or string, handle accordingly
+        if (jerry_value_is_array(argv[1])) {
```

Comment:
I think it is far easier for the user to be able to pass an array and not be burdened with creating the Buffer object themselves.  And yes, I was following the example of how it would be used in JS.